hail/python/hailtop/aiotools/fs.py

```diff
@@ -540,26 +595,29 @@ async def copy_as_dir(self, worker_pool: AsyncWorkerPool, source_report: SourceR
         if full_dest_type == AsyncFS.FILE:
             raise NotADirectoryError(full_dest)
 
-        async with WaitableSharedPool(worker_pool) as pool:
-            async for srcentry in srcentries:
-                srcfile = srcentry.url_maybe_trailing_slash()
-                assert srcfile.startswith(src)
+        async def copy_source(srcentry):
+            srcfile = srcentry.url_maybe_trailing_slash()
+            assert srcfile.startswith(src)
 
-                # skip files with empty names
-                if srcfile.endswith('/'):
-                    continue
+            # skip files with empty names
+            if srcfile.endswith('/'):
+                return
+
+            relsrcfile = srcfile[len(src):]
+            assert not relsrcfile.startswith('/')
 
-                relsrcfile = srcfile[len(src):]
-                assert not relsrcfile.startswith('/')
+            await self._copy_file_multi_part(sema, source_report, srcfile, await srcentry.status(), url_join(full_dest, relsrcfile), return_exceptions)
 
-                await pool.call(self._copy_file, source_report, srcfile, url_join(full_dest, relsrcfile))
+        await bounded_gather2(sema, *[
```

Comment:
So I believe this is correct because if return_exceptions=True, then _copy_file_multi_part (via copy_source) will never through because they will be caught and saved in the report, which means there will never be anything to cancel.  If return_exceptions=False, then we do want to cancel.